clients/sfizz_render.cpp

```diff
@@ -0,0 +1,208 @@
+
+#include <sndfile.hh>
+#include "sfizz/Synth.h"
+#include "sfizz/MathHelpers.h"
+#include "sfizz/SfzHelpers.h"
+#include "sfizz/SIMDHelpers.h"
+#include "MidiHelpers.h"
+#include "cxxopts.hpp"
+#include <fmidi/fmidi.h>
+#include <iostream>
+
+#define LOG_ERROR(ostream) std::cerr  << ostream << '\n'
+#define LOG_INFO(ostream) if (verbose) { std::cout << ostream << '\n'; }
+#define ERROR_IF(check, ostream) if ((check)) { LOG_ERROR(ostream); std::exit(-1); }
+
+struct CallbackData {
+    sfz::Synth& synth;
+    int trackIdx;
+    double lastCallTime;
+    double blockSizeInSeconds;
+    bool finished;
+};
+
+void midiCallback(const fmidi_event_t * event, void * cbdata)
+{
+    auto data = reinterpret_cast<CallbackData*>(cbdata);
+
+    if (event->type != fmidi_event_type::fmidi_event_message)
+        return;
+
+    switch (midi::status(event->data[0])) {
+        case midi::noteOff:
+            data->synth.noteOff(event->delta, event->data[1], event->data[2]);
```

Comment:
The field `event->delta` is not relevant to sequencing based on delta times, it's a raw MIDI delta value (related to tempo).

For now, the way to go is to force delta as 0, and process in small buffer sizes. (in the order of 5ms)

Ideally the library should pass the player a timestamp, which right now is not the case.

Reply:
In fact, a much better idea, after thinking better about it:

Keep the block size large, like it is.
Tick the midi player by 1/sampleRate for each frame, and increment a counter each time in the callback context, that starts at 0 for each block.
Use this counter as MIDI delay to sfizz events.